### PR TITLE
allow passing --file-reporter option to test running refs #69425

### LINKED ISSUE #69425: Flutter Test should support multiple reporters
## Use case

When running `flutter test` on a CI, output is very verbose and hard to follow:

image
(this goes on for 200+ lines)

The current alternative is to output to json with `flutter test --machine`. This is even more verbose, but can be converted to JUnit XML so we can use a specialized test visualization tools, such as Bitrise Test Reports.

This does the job, but requires a lot of configuration and there is currently no way of having two reporters at the same time.
If I want both CLI output and JSON, I need to run tests twice.
I want to be able to output to any format while still having a console output (customizable as well).

## Proposal

Flutter should have a `--reporter` CLI option, inspired by [Mocha Reporters](https://mochajs.org/#reporters) or [Jest Reporters.](https://jestjs.io/docs/en/cli#--reporters)

```bash
# Current output if not in CI environment
flutter test

# Simple, CI friendly output
flutter test --reporter spec

# Minimal output with number of passes AND json output to test_results.json file
flutter test --reporter min --reporter json --reporter-option output=test_results.json

# JUnit XML report out of the box
flutter test --reporter xunit  --reporter-option output=test_results.xml
```

By default Flutter will use the current reporter. If env `CI=true`, flutter will use `spec` automatically.

This is an idea we can take from the very well established and mature JS ecosystem.

### PR DESCRIPTION
**Pull request by mateusfccp**
Allow passing `--file-reporter` option to test runner so that both machine and human readable output can be observed in a CI setting.

Supersedes #111969.

Fixes #69425.

**Note:** I did not made the initial commit. It was made entirely by @daniel-v, so all credits goes to him. I rebased his patch and I'm going to continue the PR process by fixing tests/checks and reacting to reviewers.

## Pre-launch Checklist
  * [x]  I read the [Contributor Guide](https://github.com/flutter/flutter/wiki/Tree-hygiene#overview) and followed the process outlined there for submitting PRs.
  * [x]  I read the [Tree Hygiene](https://github.com/flutter/flutter/wiki/Tree-hygiene) wiki page, which explains my responsibilities.
  * [x]  I read and followed the [Flutter Style Guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo), including [Features we expect every widget to implement](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement).
  * [x]  I signed the [CLA](https://cla.developers.google.com/).
  * [x]  I listed at least one issue that this PR fixes in the description above.
  * [x]  I updated/added relevant documentation (doc comments with `///`).
  * [x]  I added new tests to check the change I am making, or this PR is [test-exempt](https://github.com/flutter/flutter/wiki/Tree-hygiene#tests).
  * [x]  All existing and new tests are passing.

### CODE REVIEW

**Comment by mateusfccp** [comment]
For some reason, the `Linux web_canvaskit_tests_5` check is locked in progress, but the dashboard says it completed: https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20web_canvaskit_tests_5/22458/overview

**Comment by christopherfujino** [comment]
@Jasguerrero can you review this one?

File: packages/flutter_tools/test/commands.shard/hermetic/test_test.dart [hunk 1]
```diff
@@ -804,6 +804,40 @@ dev_dependencies:
...
+
+      final TestCommand testCommand = TestCommand(testRunner: testRunner);
+      final CommandRunner<void> commandRunner = createTestCommandRunner(testCommand);
+
+      await commandRunner.run(const <String>[
+        'test',
+        '--no-pub',
+        '--file-reporter=json:out.jsonl'
+      ]);
+      expect(testRunner.lastFileReporterValue, 'json:out.jsonl');
```
Comment (Jasguerrero) [inline packages/flutter_tools/test/commands.shard/hermetic/test_test.dart:835]: Can we also add an expect that something was written to out.jsonl

Comment (mateusfccp) [reply packages/flutter_tools/test/commands.shard/hermetic/test_test.dart:835]: Do you want a test that checks if anything was written or do you want to check the specific output?

I feel the later is a little brittle, as it depends on the output of the `dart test` tool.

Comment (mateusfccp) [reply packages/flutter_tools/test/commands.shard/hermetic/test_test.dart:835]: @Jasguerrero I may be wrong, but while exploring the code to implement this request I noted that in the tests we actually don't run the test command, as we use a `FakeFlutterTestRunner` instead of a regular `_FlutterTestRunnerImpl`.

Testing this case would required an integration test. Is this what you mean?

Comment (Jasguerrero) [reply packages/flutter_tools/test/commands.shard/hermetic/test_test.dart:836]: > Do you want a test that checks if anything was written or do you want to check the specific output?

It was more like a sanity check that the file was created with something in it regardless the content (as you mentioned `dart test` should handle the content)

> Testing this case would required an integration test. Is this what you mean?

Good point. Looking at the code you could leverage https://github.com/flutter/flutter/blob/3.3/packages/flutter_tools/test/integration.shard/test_test.dart#L167 the problem with those tests is that it uses the local file system (then asserting on the stdout), writing a file to it is not the best practice. For now I think the current tests in this PR are good enough.

**Review by Jasguerrero — APPROVED** [review]
LGTM

**Review by andrewkolos — APPROVED** [review]
LGTM
